lasio/writer.py

```diff
@@ -236,19 +239,66 @@ def format_data_section_line(n, fmt, l=len_numeric_field, spacing_chars=" "):
             result = value
         return spacing_chars + result
 
+    def get_column_fmt(j):
+        """Get format string for column *j*."""
+        if j in column_fmt:
+            return column_fmt[j]
+        else:
+            return fmt
+
+    def get_left_spacing(j):
+        """Get left-hand-side spcaing for column *j*."""
+        if j == 0:
+            left_spacing = lhs_spacer
+        else:
+            left_spacing = spacer
+        return left_spacing
+
+    # Place curve mnemonics in the "~A" line.
+    if mnemonics_header:
+        # Calculate width of numeric values from the first line,
+        header_col_widths = []
+        for j in range(ncols):
+            col_fmt = get_column_fmt(j)
+            left_spacing = get_left_spacing(j)
+            data_value = format_data_section_line(
+                data_arr[0, j], col_fmt, spacing_chars=left_spacing
+            )
+            header_col_widths.append(len(data_value))
+
+        # Construct mnemonics for header line.
+        header_values = []
+        for j, curve in enumerate(las.curves):
+            col_width = header_col_widths[j]
+            if len(curve.mnemonic) > (col_width - 1):
+                width = len(curve.mnemonic) + 1
+            else:
+                width = col_width
+            value = curve.mnemonic.rjust(width)
+            header_values.append(value)
+
+        # Add data section header prefix
+        data_section_header += " "
+        if len(header_values):
+            hv = header_values[0]
+            for k in range(len(data_section_header)):
+                if k < len(hv):
+                    if hv[0] == " ":
+                        hv = hv[1:]
+            header_values = [hv] + header_values[1:]
+
```

Comment:
Not a change request for this review but after this change merges, maybe look at moving the code that generates the `data_section_header` and `header_values` to a separate function.  The benefit is that then the main flow of the `if mnemonics_header:... else...` will be clearer.   However, there might variable scope issues that make moving the code complicated...